Sisyphus v2

roydahan opened the pull request:
## PR pre-checks (self review)
<!--- PR should be created as Draft, when CI finished and relevant checkboxes selected, add reviewers and then click on "Ready for review" button.-->
<!--- Put an `x` in all the boxes that apply or create PR and then click on all relevant checkboxes: -->
- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [x] I added the relevant `backport` labels
- [x] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [x] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [x] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

roydahan commented:
The code itself is ready for review, but I'm going to change one longevity to use it and haven't tested it.
I couldn't run the unittests on my laptop, so It may contain issues.
I'm working on a way to run it.

roydahan commented:
Test is running here: https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/roy/job/longevity-200gb-48h-sisyphus/2/console

fgelcer commented on sdcm/nemesis.py:
```diff
@@ -358,6 +359,25 @@ def get_list_of_methods_by_flags(
+                r'self\.(?P<method_name>disrupt_[A-Za-z_]+?)\(.*\)', inspect.getsource(subclass), flags=re.MULTILINE)
+            if method_name:
+                disrupt_methods_list.append(method_name.group('method_name'))
+        self.log.debug("list of matching disrupions: {}".format(disrupt_methods_list))
```
change to `self.log.debug("list of matching disrupions: %s", disrupt_methods_list)`

roydahan replied:
Old style %-formatting isn't better or preferred over string formatting (if any, the old style should be removed).
I would have used f-format if not the old discussion and pylint idiotic errors.

bentsi replied:
@roydahan you should do as Fabio suggested.
https://stackoverflow.com/questions/34619790/pylint-message-logging-format-interpolation

roydahan replied:
Not true anymore.
1. I don't get pylint errors for this.
2. Even If I had, I would change the pylintrc to "new" instead of "old" which doesn't have any real benefit.
3. Not only that, since pylint 2.4, we can also use fstrings if we want for logging and won't get this warning that doesn't matter in 99% of the times.

If you want to go into a debate, you're welcome to read issue 2395 in pylint.
(I don't want to link it so it won't link this PR there).

fgelcer commented on sdcm/nemesis.py:
```diff
@@ -1101,10 +1127,35 @@ def execute_disrupt_method(self, disrupt_method):
+        disruptions = disruptions * nemesis_multiply_factor
+
+        self.disruptions_list.extend(disruptions)
+        self.log.debug("This is the list of callable disruptions {}".format(self.disruptions_list))
```
```suggestion
        self.log.debug("This is the list of callable disruptions %s", self.disruptions_list)
```

bentsi replied:
same here, use lazy interpolation

roydahan replied:
not needed.

fgelcer commented on sdcm/sct_config.py:
```diff
@@ -1205,6 +1205,20 @@ class SCTConfiguration(dict):
+             type=str_or_list,
+             help="""nemesis_filter is an INCLUDE filter. 
+             it filters IN all the nemesis that has one of the flags in the list set to true.
+             (IOW filters out all nemesis that doesn't have this flag set to true)
```
what is `IOW`?

roydahan replied:
"In other words"

fgelcer commented on sdcm/sct_config.py:
```diff
@@ -1205,6 +1205,20 @@ class SCTConfiguration(dict):
+             (IOW filters out all nemesis that doesn't have this flag set to true)
+             If a flag doesn't exist, all nemesis will be included.
+             When more than one flag is in the list, the list is the intersection of all the nemesis that has each flag 
+             (IOW it "AND" operator). """),
```
i'm not sure i understand this information inside the parenthesis

roydahan replied:
It means that it's "AND" operator and not "OR".
I'm missing there "'s", I'll push a fix for it when fixing the unit tests.

bentsi replied:
Let's be more explicit in the explanation: 
```In the other words logical AND will be used between the filter flags"

roydahan replied:
I don't see anything in the blank row you have below.
What should be more explicit?

fgelcer commented on sdcm/sct_config.py:
```diff
@@ -1205,6 +1205,20 @@ class SCTConfiguration(dict):
+
+        dict(name="nemesis_multiply_factor", env="SCT_NEMESIS_MULTIPLY_FACTOR",
+             type=int,
+             help="Multiply the list of nemesis to execute by the specified factor"),
```
is it kind of a repetition factor?

roydahan replied:
Yes, it's the number of time a nemesis will appear in the list of nemesis.

roydahan commented on defaults/test_default.yaml:
```diff
@@ -172,3 +172,5 @@ data_volume_disk_size: 0
 install_mode: 'repo'  # install from scylla_repo
 nosqlbench_image: 'scylladb/hydra-loaders:nosqlbench-4.15.49'
+
+nemesis_multiply_factor: 1
```
See sct_config.

roydahan commented:
In the next PR after this one merged, I'm going to do the following changes in all longevities:
1. Stop using ChaosMonkey and Moving to SisyphusMonkey (maybe we will change the name to ChoasMonkey.v2).
2. In master, the longevities will continue as is without being "fixed" to a specific seed.
3. In OSS releases the "seed" will be fixed so we can track regressions easier.
4. A new default multiply_factor will be decided.

bentsi commented on sdcm/sct_config.py:
```diff
@@ -1205,6 +1205,20 @@ class SCTConfiguration(dict):
              type=int,
              help="Number of iops for ebs type io2|io3|gp3"),
+
+        dict(name="nemesis_filter", env="SCT_NEMESIS_FILTER",
```
I would call it `nemesis_include_filter`

roydahan replied:
At first I called it include_list, but I think it's more confusing.
Think about it, usually when you filter you "select" the things you would like to see.

bentsi commented on sdcm/nemesis.py:
```diff
@@ -387,7 +407,13 @@ def _get_subclasses_from_list(
+                try:
+                    attr = getattr(nemesis, filter_name)
+                except AttributeError:
+                    LOGGER.warning("The required nemesis flag was not found")
```
```suggestion
                    LOGGER.warning("The required nemesis flag '%s' was not found", filter_name)
```

bentsi commented on test-cases/longevity/longevity-200GB-48h-verifier-LimitedMonkey-tls.yaml:
```diff
@@ -1,5 +1,5 @@
 test_duration: 3100
-prepare_write_cmd: "cassandra-stress write cl=ALL n=200200300  -schema 'replication(factor=3) compaction(strategy=LeveledCompactionStrategy)' -port jmx=6868 -mode cql3 native -rate threads=1000 -col 'size=FIXED(1024) n=FIXED(1)' -pop seq=1..200200300 -log interval=15"
+prepare_write_cmd: "cassandra-stress write cl=ALL n=200200  -schema 'replication(factor=3) compaction(strategy=LeveledCompactionStrategy)' -port jmx=6868 -mode cql3 native -rate threads=1000 -col 'size=FIXED(1024) n=FIXED(1)' -pop seq=1..200200 -log interval=15"
```
is this intended to be in? or you had used it for testing?

roydahan replied:
It's just for testing, I'll remove it in next commits.

roydahan commented:
@bentsi , @fgelcer please review this one.
I would like to merge it ASAP so we can start using it.

I tested it only on one longevity before addressing the comments and it worked.
We will test it completely as part of 4.6 if we merge it ASAP.

fgelcer commented on test-cases/longevity/longevity-100gb-4h.yaml:
```diff
@@ -12,7 +12,7 @@ n_monitor_nodes: 1
 instance_type_db: 'i3.4xlarge'
 
-nemesis_class_name: 'ChaosMonkey'
+nemesis_class_name: 'SisyphusMonkey'
```
is it on purpose it doesn't have a `nemesis_seed` set?

roydahan replied:
No, I'll fix it.

roydahan replied:
fixed

fgelcer commented on test-cases/longevity/longevity-1TB-5days-authorization-and-tls-ssl.yaml:
```diff
@@ -29,10 +30,10 @@ gce_instance_type_loader: 'c2-standard-16'
 
 
-nemesis_class_name: 'ChaosMonkey'
+nemesis_class_name: 'SisyphusMonkey'
```
same question about `nemesis_seed` here

roydahan replied:
fixing.

roydahan replied:
fixed.

fgelcer commented on test-cases/longevity/longevity-100GB-48h-cloud-CloudLimitedChaosMonkey-tls.yaml:
```diff
@@ -15,7 +15,8 @@ post_behavior_db_nodes: "destroy"
 
-nemesis_class_name: 'LimitedChaosMonkey'
+nemesis_class_name: 'SisyphusMonkey'
+nemesis_include_filter: ['limited']
```
have you tested this `nemesis_include_filter`?

roydahan replied:
Yes, in the previous iteration, see above comment.

roydahan replied:
On this one it's on purpose, because I want it to run randomly for now.

fgelcer reviewed (dismissed):
few questions, but LGTM

bentsi commented on sdcm/nemesis.py:
```diff
@@ -1101,10 +1127,36 @@ def execute_disrupt_method(self, disrupt_method):
 
-    def get_all_disrupt_methods(self):
-        all_disruptions = [attr[1] for attr in inspect.getmembers(self)
+    def build_list_of_disruptions_to_execute(self, nemesis_include_filter=None, nemesis_multiply_factor=1):
```
nemesis_multiply_factor is not used

roydahan replied:
It's explained below.
It's there for unit test purpose, but I took them out for now, till I decide what to do with them.

bentsi replied:
I don't understand how it is going to be used in unit-test if you override it on line 1155...

bentsi reviewed (commented):
we can merge it if it is urgent, but need to fix the issue with overridden `nemesis_multiply_factor` argument in build_list_of_disruptions_to_execute

fgelcer reviewed (approved):
LGTM

fgelcer reviewed (approved):
LGTM

bentsi commented:
autopep8 failed
![image](https://user-images.githubusercontent.com/5269241/146393962-89999555-b668-4bdb-83b7-3a3dad98304d.png)

fgelcer commented:
> autopep8 failed ![image](https://user-images.githubusercontent.com/5269241/146393962-89999555-b668-4bdb-83b7-3a3dad98304d.png)

fixed right now

fgelcer reviewed (approved):
LGTM
